Async support, configurable staging path, tar file support

3coins opened the pull request:
## Updates

1. Slight optimization on `get_staging_paths` method to avoid calling `add_outputs`
2. Added docs for `get_staging_paths`
3. Made `staging_path` a configurable attribute
4. Updated `OutputFilesManager` to support async method calls to Scheduler apis
5. Updated `Downloader` to extract output files from tar and compressed tar files, tested this with staging path set to both local file and S3

3coins commented:
Updated based on discussion with @ellisonbg. Added some additional classes to allow outputs export in compressed tar format.

ellisonbg commented on jupyter_scheduler/handlers.py:
```diff
@@ -272,7 +272,7 @@ def output_files_manager(self):
     async def get(self, job_id):
         redownload = self.get_query_argument("redownload", False)
-        self.output_files_manager.copy_from_staging(job_id=job_id, redownload=redownload)
+        await self.output_files_manager.copy_from_staging(job_id=job_id, redownload=redownload)
```
Does this solve the frontend issue of knowing when to show the links?

3coins replied:
No, this won't fix the front end issue.

ellisonbg commented on jupyter_scheduler/executors.py:
```diff
@@ -126,13 +128,16 @@ def execute(self):
-                f.write(output)
+        try:
+            ep.preprocess(nb)
+        except CellExecutionError as e:
```
Looks like you are also solving the issue related to cell executions errors being treated as job failures.

3coins commented:
Updates:
1. Refactored the `create_output_filename` to a separate function in `BaseScheduler`. This helps simplifying getting the output filenames from the `OutputFilesManager`. The advantage is that this will support having different filenames in staging vs local workspace.
2. Refactored the `download*` functions in `Downloader` to use a common function to extract input/output file paths.

dlqqq commented on jupyter_scheduler/scheduler.py:
```diff
@@ -470,12 +530,41 @@ def list_job_definitions(self, query: ListJobDefinitionsQuery) -> ListJobDefinit
     def get_staging_paths(self, job_id: str) -> Dict[str, str]:
-        model = self.get_job(job_id)
+        model = None
+        with self.db_session() as session:
+            model = session.get(Job, job_id)
```
This is actually a very relevant performance concern, and it's good that you caught this. When https://github.com/jupyter-server/jupyter-scheduler/pull/175 gets merged, building that model will require an expensive `file_id_manager.get_path()` call. In summary, the issue is that there are many scenarios in which we want to just fetch the job **record** instead of the `DescribeJob` model, because building that model involves additional compute load to bind attributes we don't need.

This is a cross-cutting concern outside of this class as well. We want to do the same in `OutputFilesManager.copy_from_staging()`.

Thus, I believe the solution here is that we add a new public method on scheduler called `get_job_record()`, which just returns the job record instead of the `DescribeJob` model. We should probably do the same for `get_job_definition_record()` just to preserve parity between job & job definition methods, and so we don't have to ask custom implementors to add another method twice.

dlqqq replied:
Thinking about this a bit more deeply, `OutputFilesManager.copy_from_staging()` shouldn't call `get_job_record()`, since the return type of this method would vary per implementation. Hence, we should just keep these as private methods and restrict their use to `Scheduler`.

dlqqq commented on jupyter_scheduler/scheduler.py:
```diff
@@ -470,12 +530,41 @@ def list_job_definitions(self, query: ListJobDefinitionsQuery) -> ListJobDefinit
+        return staging_paths
+
+
+class ArchiveDownloadingScheduler(Scheduler):
+    """Scheduler that adds archive path to staging paths.
+
+    Notes
+    -----
+    Should be used in tandem with :class:`~jupyter_scheduler.executors.ArchiveExportingExecutingManager`
+    during jupyter server start.
+    """
```
This class should set its own default for the `execution_manager_class` trait. That way users don't need to manually don't need to manually specify the correct execution manager class in tandem.

3coins replied:
Good idea!

dlqqq commented on jupyter_scheduler/scheduler.py:
```diff
@@ -470,12 +530,41 @@ def list_job_definitions(self, query: ListJobDefinitionsQuery) -> ListJobDefinit
+        return staging_paths
+
+
+class ArchiveDownloadingScheduler(Scheduler):
```
Rename this to `ArchivingScheduler` for brevity?

dlqqq reviewed (changes requested):
This looks good overall, aside from a few minor concerns. Some thoughts on the high-level design:

1. Perhaps we add an `archive: bool` trait on the `SchedulerApp` extension that automatically sets the correct `scheduler_class`? e.g.

```py
@default('scheduler_class')
def _default_scheduler_class():
  return "jupyter_scheduler.scheduler.ArchiveDownloadingScheduler" if self.archive else "jupyter_scheduler.scheduler.Scheduler"
```

2. I would like to see at least one end-to-end unit test for using `ArchivingScheduler` and `ArchivingExecutionManager` (e.g. create a job, generate the output files, verify they are there). But given that I don't think there are any at the moment, we should track this in a separate issue to avoid blocking this PR.

3coins commented on jupyter_scheduler/scheduler.py:
```diff
@@ -470,12 +530,41 @@ def list_job_definitions(self, query: ListJobDefinitionsQuery) -> ListJobDefinit
+                filename = create_output_filename(model.input_uri, model.create_time, output_format)
+                staging_paths[output_format] = os.path.join(filename)
+
+            output_format = "tar.gz"
+            filename = create_output_filename(model.input_uri, model.create_time, output_format)
+            staging_paths[output_format] = os.path.join(self.staging_path, job_id, filename)
```
tar.gz has different staging_path that the files. In the archived case, tar.gz points to the actual full path to the archive, but the files refer to the path inside the tar file.

3coins commented on jupyter_scheduler/scheduler.py:
```diff
@@ -470,12 +530,41 @@ def list_job_definitions(self, query: ListJobDefinitionsQuery) -> ListJobDefinit
+        if model:
+            for output_format in model.output_formats:
+                filename = create_output_filename(model.input_uri, model.create_time, output_format)
+                staging_paths[output_format] = os.path.join(filename)
```
It should just be filename, will remove `os.path.join`.

dlqqq commented:
Looks good! Unit tests can be addressed later since there are plenty of other areas in the backend that need them.
